R/data-gather-clean.R

```diff
@@ -0,0 +1,121 @@
+                            section_lookup_table, variable_lookup_table) {
+  if (is.null(statuses)) {
+    return(NULL)
+  }
+  submissions <- purrr::map(statuses, function(x) {
+    synapseforms::download_all_and_get_table(
+      syn = syn,
+      state_filter = x,
+      group = group
+    )
+  }) %>%
+    purrr::compact() %>% # Removes NAs
+    purrr::reduce(dplyr::full_join, by = "variables")
+
+  if (!is.null(submissions)) {
+    submissions <- make_clean_table(
+      submissions,
+      section_lookup_table,
+      variable_lookup_table
+    )
+  }
+  submissions
+}
+
+#' Make clean submissions table
+#'
+#' Make submissions table clean by fixing experiment
+#' numbering and adding in user-friendly step and label
+#' columns instead of json variable styled section and
+#' variable columns.
+#'
+#' @keywords internal
+#' @inheritParams get_submissions
+#' @inheritParams clean_experiment_variables
+make_clean_table <- function(data, section_lookup_table,
+                             variable_lookup_table) {
+  data <- synapseforms::make_tidier_table(data)
+  data <- dplyr::full_join(data, section_lookup_table, by = "section")
+  data <- clean_experiment_variables(data)
+  data <- dplyr::full_join(data, variable_lookup_table, by = "variable")
+  data <- data[!is.na(data$response), ]
+  # Some labels/steps will be blank if form changes or not mapped
+  na_labels <- which(is.na(data$label))
+  data$label[na_labels] <- data$variable[na_labels]
+  na_steps <- which(is.na(data$step))
+  data$step[na_steps] <- data$section[na_steps]
+  # For now, it would be easier to put a "fake" column for
+  # submission name until we get user-friendly names.
+  # Currently considering form_data_id the name.
+  data <- tibble::add_column(data, submission = data$form_data_id)
+  data <- data[c(
+    "form_data_id",
+    "submission",
+    "step",
+    "label",
+    "response"
+  )]
+  data
+}
+
+#' Clean experiment variables
+#'
+#' Multiple experiments lead to `variable`s with names of the
+#' form "age_range1", "age_range2". This will remove the number at
+#' the end of `variable` and append to the `step` name.
+#' Note that this only happens if the number is > 0. This is due
+#' to having one known `variable`, ID50, that ends in a number,
+#' but is not related to the multiple experiment issue.
+#'
+#' @param data The submission data in the form given by
+#'   [synapseforms::make_tidier_table].
+clean_experiment_variables <- function(data) {
+  # Basic section can have multiple routes, but if split first,
+  # these ones won't get messed up due to not being tertiary to section.
+  data <- tidyr::separate(
+    data,
+    col = "variable",
+    sep = "[.]",
+    into = c("main_variable", "sub_variable"),
+    remove = FALSE,
+    extra = "merge",
+    fill = "right"
+  )
+  num_list <- purrr::map(
+    data$sub_variable,
+    function(x) {
+      suppressWarnings(as.numeric(substr(x, nchar(x), nchar(x))))
+    }
+  )
```

Comment:
This will fail if there are 10 or more experiments; maybe not likely, but we should still handle that case. Maybe let's proceed but open an issue to fix it afterward.

Reply:
I think this fix wouldn't take too long. I can probably add it to this branch.

I believe my reasoning for doing it this way was:
1. it seems like a low chance of 10+ experiments for a single section, and
2. I was nervous about tying any portion of the code to a specific variable (i.e. checking if it's ld50 and making that the only edge case). This is because the JSON schema for the forms could potentially change in the future such that ld50 isn't the only variable to end in a number.

However, doing it this way has more issues than just it messing up with 10+ experiments. Again, the JSON schema could change to have a variable that ends in a digit between 1-9 and it would mess up the table, as well.

Reply:
Alrighty, I pushed a commit that "mostly" fixes this issue. It should work fine for 10+ experiments.

One issue leftover is that there is a variable that does not fit the format of other variables due to being able to choose more than one. If one experiment, let's say ld50, has multiple routes of administration for the drug, then the JSON variables will be ld50.experiments.route1, ld50.experiments.route2, ld50.experiments.route3, etc. The way the code is written now, this turns into 3 ld50 experiments (ld50 [1], ld50 [2], ld50 [3]), where all the variables except route2 and route3 will be under ld50 [1]. ld50 [2] will have a single row for route2 and ld50 [3] will have a single row for route [3]. This is the only variable I have found so far that has this problem. I'll open an issue for that specific case.